backend/src/controller/chatController.ts

```diff
@@ -123,17 +123,17 @@ async function handleHigherLevelChat(
 	chatHistory: ChatHistoryMessage[],
 	defences: Defence[]
 ): Promise<LevelHandlerResponse> {
-	let updatedChatHistory = [...chatHistory];
-
 	// transform the message according to active defences
 	const transformedMessage = transformMessage(message, defences);
-	const chatHistoryUserMessages = createNewUserMessages(
+	const newUserMessagesForChatHistory = createNewUserMessages(
 		message,
 		transformedMessage ? combineTransformedMessage(transformedMessage) : null
 	);
-	chatHistoryUserMessages.forEach((message) => {
-		updatedChatHistory = pushMessageToHistory(updatedChatHistory, message);
-	});
+
+	const chatHistoryWithNewUserMessages = newUserMessagesForChatHistory.reduce(
```

Comment:
You are creating `const newUserMessagesForChatHistory` only to use it in this reduce call. Be more economic with your code where it doesn't reduce readability:

```
const chatHistoryWithNewUserMessages =
  createNewUserMessages(message, transformedMessage)
    .reduce(pushMessageToHistory, chatHistory);
```